### DISCUSSION TITLE
Randomizing multiple images for one trial

### DISCUSSION THREAD

**Discussion by BobRoss048** [jspsych/jsPsych#1054] [Q&A] [answered]
Hello,

I am using JsPsych for a simple reaction task.  Participants see an image and indicate what they see in a single trial run (this is important for my experiment).  

My question is, can I give the single trial multiple images to choose from, but have it only choose one for a single trial, or is it easier for me to just create multiple conditions or trials if I want to pull this off?  

Essentially, I have it set up so that one of two images needs to appear, but only once and only in the single trial.  So far, if I had two images, a minimum of two trials occur to get through both images.  

Thank you.

**Comment by becky-gilbert** [comment]
It sounds like you want to loop over a set of images, and maybe have them occur in random order. Is that right? If so, then it sounds like you want to use [timeline variables](https://www.jspsych.org/overview/timeline/#timeline-variables). The example from the documentation might be very similar to what you want. Something like this:
```javascript
var srt_procedure = {
    timeline: [
        {
            type: 'html-keyboard-response',
            stimulus: '+',
            choices: jsPsych.NO_KEYS,
            trial_duration: 500
        },
        {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('stim'),
            choices: ['space'],
            trial_duration: 3000,
            response_ends_trial: true
        }
    ],
    timeline_variables: [
        { stim: 'image-1.jpg' },
        { stim: 'image-2.jpg' },
        { stim: 'image-3.jpg' },
        { stim: 'image-4.jpg' }
    ],
    randomize_order: true
}
```
Is that what you wanted? Apologies if I misunderstood.

**Reply by BobRoss048** [reply]
Hello Becky,

Thank you for your response to my question.  It might be the answer I'm looking for.  My question is, will this pick a single image among the four (at random) and end the trial after that image?  My goal is for it not to go through all four images (in your example) and just pick a single image every time.  Once the participant indicates the image, the experiment ends.  

I hope this clears it up a little bit.  Sorry, I'm new to this and so my explanation is a bit rough.

**Reply by becky-gilbert** [reply]
Sorry, I misunderstood! I'll reply in a separate comment.

**Comment by becky-gilbert** [comment, marked as answer]
You can create an array of your images to choose from, and then sample randomly from it using jsPsych's [sampleWithoutReplacement](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationsamplewithreplacement) function. You could either do this when the page first loads, like this:
```javascript
var all_images = ['image-1.jpg', 'image-2.jpg', 'image-3.jpg', 'image-4.jpg'];
var stim = jsPsych.randomization.sampleWithoutReplacement(all_images, 1)[0];
var image_trial = {
  type: 'image-keyboard-response',
  stimulus: stim,
  choices: ['space'],
  trial_duration: 3000,
  response_ends_trial: true
}
```

Or, if you want the random selection to happen dynamically on each trial, you could make your [`stimulus` parameter into a function](https://www.jspsych.org/overview/trial/#dynamic-parameters). Then, right before the trial is about to run, the stimulus function will run and return the value that should be used on that trial, e.g.:
```javascript
var all_images = ['image-1.jpg', 'image-2.jpg', 'image-3.jpg', 'image-4.jpg'];
var image_trial = {
  type: 'image-keyboard-response',
  stimulus: function() {
    return jsPsych.randomization.sampleWithoutReplacement(all_images, 1)[0];
  },
  choices: ['space'],
  trial_duration: 3000,
  response_ends_trial: true
}
```

**Reply by jodeleeuw** [reply]
You can also use the sample parameter for timeline variables. On my phone right now so hard for me to leave an example, but it is in the docs.

**Reply by becky-gilbert** [reply]
Ah, I totally forgot that you can also do [sampling with timeline variables](https://www.jspsych.org/overview/timeline/#sampling-methods)! This might be a nicer solution. If you're only selecting one image/trial from the array, then you could use either the [with replacement](https://www.jspsych.org/overview/timeline/#sampling-with-replacement) or [without replacement](https://www.jspsych.org/overview/timeline/#sampling-without-replacement) sampling method, e.g.:
```javascript
var srt_procedure = {
    timeline: [
        {
            type: 'html-keyboard-response',
            stimulus: '+',
            choices: jsPsych.NO_KEYS,
            trial_duration: 500
        },
        {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('stim'),
            choices: ['space'],
            trial_duration: 3000,
            response_ends_trial: true
        }
    ],
    timeline_variables: [
        { stim: 'image-1.jpg' },
        { stim: 'image-2.jpg' },
        { stim: 'image-3.jpg' },
        { stim: 'image-4.jpg' }
    ],
    sample: {
        type: 'with-replacement',
        size: 1, // randomly choose one object from the timeline_variables array
    }
}
```

**Reply by BobRoss048** [reply]
I'll give these examples a shot to see how they do for my situation.  

Thank you all for all your assistance.
